library/connect.c

```diff
@@ -709,39 +711,47 @@ static int send_crypto_header(ziti_connection conn) {
     return ZITI_OK;
 }
 
-static void on_flush(uv_check_t *fl) {
+static void on_flush(uv_idle_t *fl) {
     ziti_connection conn = fl->data;
     if (conn == NULL) {
         uv_close((uv_handle_t *) fl, (uv_close_cb) free);
         return;
     }
-    uv_check_stop(fl);
-    flush_to_service(conn);
-    flush_to_client(conn);
+
+    bool more_to_client = flush_to_client(conn);
+    bool more_to_service = flush_to_service(conn);
+
+    if (!more_to_client && !more_to_service) {
+        CONN_LOG(TRACE, "stopping flusher");
+        uv_idle_stop(fl);
+    }
 }
 
 static void flush_connection (ziti_connection conn) {
     if (conn->flusher) {
-        uv_check_start(conn->flusher, on_flush);
+        CONN_LOG(TRACE, "starting flusher");
+        uv_idle_start(conn->flusher, on_flush);
     }
 }
 
-static void flush_to_service(ziti_connection conn) {
+static bool flush_to_service(ziti_connection conn) {
 
-    if (conn->state == Connecting) {
-        flush_connection(conn);
-    } else {
+    if (conn->state == Connected) {
+        int count = 0;
         while (!TAILQ_EMPTY(&conn->wreqs)) {
             struct ziti_write_req_s *req = TAILQ_FIRST(&conn->wreqs);
             TAILQ_REMOVE(&conn->wreqs, req, _next);
 
             conn->write_reqs++;
             ziti_write_req(req);
         }
+        CONN_LOG(TRACE, "flushed %d messages", count);
     }
+
+    return !TAILQ_EMPTY(&conn->wreqs);
 }
 
-static void flush_to_client(ziti_connection conn) {
+static bool flush_to_client(ziti_connection conn) {
```

Comment:
nit. If I'm not mistaken the return value indicates whether or not data is still pending, which is the opposite of how I naturally read the signature - whether or not the flush completed. IMO it would be more clear if the function returned the number of bytes remaining.